contracts/programs/gummyshop/src/lib.rs

```diff
@@ -0,0 +1,356 @@
+        Ok(())
+    }
+
+    pub fn list(
+        ctx: Context<AddModifyListing>,
+        price: u64,
+        data_hash: [u8; 32],
+        creator_hash: [u8; 32],
+        nonce: u128,
+        index: u32,
+        root: [u8; 32]
+    ) -> Result<()> {
+        modify_compressed_nft_delegate(
+            &ctx.accounts.owner,
+            &ctx.accounts.former_delegate,
+            &ctx.accounts.new_delegate,
+            &ctx.accounts.bubblegum_authority,
+            &ctx.accounts.gummyroll,
+            &ctx.accounts.merkle_slab,
+            &ctx.accounts.bubblegum,
+            &ctx.accounts.system_program,
+            data_hash,
+            creator_hash,
+            nonce,
+            index,
+            root
+        )?;
+        Ok(())
+    }
+
+    pub fn modify_listing(
+        ctx: Context<AddModifyListing>,
+        price: u64,
+        data_hash: [u8; 32],
+        creator_hash: [u8; 32],
+        nonce: u128,
+        index: u32,
+        root: [u8; 32]
+    ) -> Result<()> {
+        modify_compressed_nft_delegate(
+            &ctx.accounts.owner,
+            &ctx.accounts.former_delegate,
+            &ctx.accounts.new_delegate,
+            &ctx.accounts.bubblegum_authority,
+            &ctx.accounts.gummyroll,
+            &ctx.accounts.merkle_slab,
+            &ctx.accounts.bubblegum,
+            &ctx.accounts.system_program,
+            data_hash,
+            creator_hash,
+            nonce,
+            index,
+            root
+        )?;
+        Ok(())
+    }
+
+    pub fn remove_listing(
+        ctx: Context<RemoveListing>,
+        data_hash: [u8; 32],
+        creator_hash: [u8; 32],
+        nonce: u128,
+        index: u32,
+        root: [u8; 32]
+    ) -> Result<()> {
+        modify_compressed_nft_delegate(
+            &ctx.accounts.owner,
+            &ctx.accounts.former_delegate,
+            &ctx.accounts.new_delegate,
+            &ctx.accounts.bubblegum_authority,
+            &ctx.accounts.gummyroll,
+            &ctx.accounts.merkle_slab,
+            &ctx.accounts.bubblegum,
+            &ctx.accounts.system_program,
+            data_hash,
+            creator_hash,
+            nonce,
+            index,
+            root
+        )?;
+        Ok(())
+    }
+
+    pub fn purchase<'info>(
+        ctx: Context<'_, '_, '_, 'info, Purchase<'info>>,
+        price: u64,
+        data_hash: [u8; 32],
+        root: [u8; 32],
+        nonce: u128,
+        index: u32,
+        creator_shares: Vec<u8>
+    ) -> Result<()> {
+        // First, payout the marketplace's royalty fee. This has to be off the top since Bubblegum mint currently
+        // sets the base creator's share to 100%.
+        // TODO(sorend): consider using bigdecimal for casting and intermediate operations to avoid loss of information for projects with high price 
+        let amount_to_pay_marketplace = (price as f64 * (ctx.accounts.marketplace_props.share as f64/100.0)) as u64;
```

Comment:
Floating point feels like it could get a little nasty. We can check this when we can get tests working

Reply:
Agree, upgraded this in the latest contract. I think it should give us more safety in general, but might not work for projects with ridiculously cheap prices.